system/syslogd/syslogd_main.c

```diff
@@ -0,0 +1,376 @@
+#endif
+          break;
+
+        case 'n':
+
+          /* Stay in foreground */
+
+#ifdef CONFIG_LIBC_EXECFUNCS
+          background = false;
+#endif
+          break;
+
+        case '?':
+          print_usage();
+          exit(EXIT_FAILURE);
+          break;
+        }
+    }
+
+    /* Run this program in the background as a spawned task if the background
+     * option was selected.
+     */
+
+#ifdef CONFIG_LIBC_EXECFUNCS
+  if (background)
+    {
+      /* Set up the arguments, which is identical to the original except with
+       * an added `-n` flag to ensure that the new process does not 'respawn'
+       */
+
+      new_argv = malloc(sizeof(char *) * (argc + 1));
+      new_argv[0] = argv[0]; /* Same program name */
+      new_argv[1] = "-n";    /* Prevent daemon from spawning another child */
+      memcpy(&new_argv[2], &argv[1], sizeof(char *) * (argc - 1));
+
+      /* Spawn the child for backgrounding now */
+
+      if (posix_spawn(&pid, argv[0], NULL, NULL, new_argv, NULL) != 0)
+        {
+          fprintf(stderr, "Failed to fork() to background process: %d\n",
+                  errno);
+          free(new_argv);
+          return EXIT_FAILURE;
+        }
+      else
+        {
+          /* We succeeded in spawning, exit now. */
+
+          free(new_argv);
+          return EXIT_SUCCESS;
+        }
+    }
+#endif /* CONFIG_LIBC_EXECFUNCS */
+
+  /* Set up client connection information */
+
+  server.sin_family = AF_INET;
+  server.sin_port = htons(CONFIG_SYSTEM_SYSLOGD_PORT);
+  server.sin_addr.s_addr = inet_addr(CONFIG_SYSTEM_SYSLOGD_ADDR);
+
+  if (server.sin_addr.s_addr == INADDR_NONE)
+    {
+      fprintf(stderr, "Invalid address '%s'\n", CONFIG_SYSTEM_SYSLOGD_ADDR);
+      return EXIT_FAILURE;
+    }
+
+  /* Create a UDP socket */
+
+  if (debugmode)
+    {
+      printf("Creating UDP socket %s:%u\n", CONFIG_SYSTEM_SYSLOGD_ADDR,
+             CONFIG_SYSTEM_SYSLOGD_PORT);
+    }
+
+  sock = socket(AF_INET, SOCK_DGRAM, 0);
+  if (sock < 0)
+    {
+      fprintf(stderr, "Couldn't create UDP socket: %d\n", errno);
+      return EXIT_FAILURE;
+    }
+
+  /* Open syslog stream */
+
+  if (debugmode)
+    {
+      printf("Opening syslog device '%s' to read entries.\n",
+             CONFIG_SYSLOG_DEVPATH);
+    }
+
+  fd = open(CONFIG_SYSLOG_DEVPATH, O_RDWR);
+  if (fd < 0)
+    {
+      fprintf(stderr, "Could not open syslog stream: %d", errno);
+      close(sock);
+      return EXIT_FAILURE;
+    }
+
+  /* Transmit syslog messages forever */
+
+  if (debugmode)
+    {
+      printf("Beginning to continuously transmit syslog entries.\n");
+    }
+
+  while (true)
+    {
+      /* Read as much data as possible into the remaining space in our buffer
+       */
+
+      bread = read(fd, &buffer[bufpos], sizeof(buffer) - bufpos);
+      if (bread < 0)
+        {
+          fprintf(stderr, "Failed to read from syslog: %d", errno);
+          continue; /* Just try again */
+        }
+
+      if (bread == 0 && bufpos == 0)
```

Comment:
we should output the buffer log and exit loop if `bread <= 0`:
```suggestion
      if (bread <= 0)
```

Reply:
since read return == 0 once, it will always 0 immediately in future, if the last message doesn't end with `\n`, syslogd will busy loop forever.